src/lib/shadow-shim-helper-rs/src/scmutex.rs

```diff
@@ -1,23 +1,139 @@
 use rkyv::{Archive, Serialize};
-use std::{
-    cell::UnsafeCell,
-    marker::PhantomData,
-    ops::Deref,
-    pin::Pin,
-    sync::atomic::{AtomicI32, AtomicU32, Ordering},
-};
+use std::{marker::PhantomData, ops::Deref, pin::Pin};
 use vasi::VirtualAddressSpaceIndependent;
 
+// When running under the loom model checker, use its sync primitives.
+// See https://docs.rs/loom/latest/loom/
+mod sync {
+    // Use std sync primitives, or loom equivalents
+    #[cfg(loom)]
+    pub use loom::{
+        sync::atomic,
+        sync::atomic::{AtomicI32, AtomicU32, Ordering},
+        sync::Arc,
+    };
+    #[cfg(not(loom))]
+    pub use std::{
+        sync::atomic,
+        sync::atomic::{AtomicI32, AtomicU32, Ordering},
+        sync::Arc,
+    };
+
+    // Map a *virtual* address to a list of Condvars. This doesn't support mapping into multiple
+    // processes, or into different virtual addresses in the same process, etc.
+    #[cfg(loom)]
+    use loom::sync::{Condvar, Mutex};
+    #[cfg(loom)]
+    use std::collections::HashMap;
+    #[cfg(loom)]
+    loom::lazy_static! {
+        pub static ref FUTEXES: Mutex<HashMap<usize, Vec<Arc<Condvar>>>> = Mutex::new(HashMap::new());
+    }
+
+    #[cfg(not(loom))]
+    pub unsafe fn futex(
+        futex_word: &AtomicI32,
+        futex_op: i32,
+        val: i32,
+        timeout: *const libc::timespec,
+        uaddr2: *mut i32,
+        val3: i32,
+    ) -> nix::Result<i64> {
+        nix::errno::Errno::result(unsafe {
+            libc::syscall(
+                libc::SYS_futex,
+                futex_word,
+                futex_op,
+                val,
+                timeout,
+                uaddr2,
+                val3,
+            )
+        })
+    }
+    #[cfg(loom)]
+    // Models the OS's futex primitive, using loom primitives.
+    // Probably not as "nondeterministic" as the real futex syscall, but have to
+    // start somewhere...
+    pub unsafe fn futex(
+        futex_word: &AtomicI32,
+        futex_op: i32,
+        val: i32,
+        _timeout: *const libc::timespec,
+        _uaddr2: *mut i32,
+        _val3: i32,
+    ) -> nix::Result<i64> {
+        match futex_op {
+            libc::FUTEX_WAIT => {
+                let mut hashmap = FUTEXES.lock().unwrap();
+                let futex_word_val = futex_word.load(Ordering::AcqRel);
```

Comment:
This was changed in the next commit.

Reply:
Yeah, whoops. I later changed this to `Relaxed`. I  don't think anything stronger is needed, since it's already synchronized through `FUTEXES.lock()`.

(This slipped by when initially sketching things out because the loom implementations don't panic on unexpected orderings the way the std equivalents do)